### PR TITLE
planner: use max correlation in heuristic row count estimation

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

For queries like
```
select id from tbl where asc_25 between '2020-05-16' and date_add('2020-05-17', interval 100 day) and asc_90 between '2020-05-16' and date_add('2020-05-17', interval 100 day) order by id limit 1;
```

column `asc_90` is more decisive for the row count estimation, so we should use max correlation if conditions contain multiple columns, instead of the previous compound correlation, otherwise, `tidb_opt_correlation_exp_factor` may have little effect to adjust the preference of index scan.

### What is changed and how it works?

- Remove duplicate extracted columns;
- Compute max correlation instead of compound correlation;
- Enable heuristic estimation by default;
- Fix bug of setting session variable for `tidb_opt_correlation_exp_factor`;

### Check List 

Tests 

 - Unit test: existing tests cover part of the change
 - Integration test: would be tested in plan stability test framework

Code changes

N/A

Side effects

N/A

Related changes

N/A

### CODE REVIEW

File: sessionctx/variable/session.go [hunk 1]
```diff
@@ -699,7 +699,7 @@ func (s *SessionVars) SetSystemVar(name string, val string) error {
 	case TiDBOptCorrelationThreshold:
 		s.CorrelationThreshold = tidbOptFloat64(val, DefOptCorrelationThreshold)
 	case TiDBOptCorrelationExpFactor:
-		s.CorrelationExpFactor = tidbOptPositiveInt32(val, DefOptCorrelationExpFactor)
+		s.CorrelationExpFactor = int(tidbOptInt64(val, DefOptCorrelationExpFactor))
```
Comment (lamxTyler) [inline sessionctx/variable/session.go]: Why change it?

File: planner/core/find_best_task.go [hunk 2]
```diff
@@ -619,33 +619,35 @@ func splitIndexFilterConditions(conditions []expression.Expression, indexColumns
...
-// single factors.
+// contains multiple columns, return nil and get the max correlation, which would be used in the heuristic estimation.
 func getMostCorrColFromExprs(exprs []expression.Expression, histColl *statistics.Table, threshold float64) (*expression.Column, float64) {
 	var cols []*expression.Column
 	cols = expression.ExtractColumnsFromExpressions(cols, exprs, nil)
 	if len(cols) == 0 {
 		return nil, 0
 	}
-	compCorr := 1.0
+	colSet := make(map[int64]interface{})
```
Comment (lamxTyler) [inline planner/core/find_best_task.go]: Use `set.Int64Set` instead?

File: sessionctx/variable/session.go (hunk 1 above)
Comment (eurekaka) [reply sessionctx/variable/session.go]: According to `tidbOptPositiveInt32`, if default value is 1, we cannot set the value to zero successfully.

Comment (eurekaka) [reply sessionctx/variable/session.go]: According to `tidbOptPositiveInt32`, if default value is 1, we cannot set the value to zero successfully.

File: planner/core/find_best_task.go [hunk 3]
```diff
@@ -804,6 +807,13 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 	if prop.ExpectedCnt < ds.stats.RowCount {
 		count, ok, corr := ds.crossEstimateRowCount(path, prop.ExpectedCnt, candidate.isMatchProp && prop.Items[0].Desc)
 		if ok {
+			// TODO: actually, before using this count as the estimated row count of table scan, we need additionally
+			// check if count < row_count(first_region | last_region), and use the larger one since we build one copTask
```
Comment (winoros) [inline planner/core/find_best_task.go]: Do we really need this todo? Coprocessor also breaks when it scans the first needed row?

Comment (eurekaka) [reply planner/core/find_best_task.go]: > Coprocessor also breaks when it scans the first needed row?

Hmm I don't understand this comment, could you please elaborate more on this? I am not saying that we modify coprocessor to return just needed rows, I mean that we should set rowCount as:
```golang
rowCount = math.Max(count, regionRowCount)
// regionRowCount is the row count of first region or last region based on ASC/DESC
```

**Pull request merged**